clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -515,20 +516,29 @@ private static int parseAcks(String acksString) {
     /**
      * Wait for cluster metadata including partitions for the given topic to be available.
      * @param topic The topic we want metadata for
+     * @param partition A specific partition expected to exist in metadata, or null if there's no preference
      * @param maxWaitMs The maximum time in ms for waiting on the metadata
      * @return The cluster containing topic metadata and the amount of time we waited in ms
      */
-    private ClusterAndWaitTime waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
+    private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long maxWaitMs) throws InterruptedException {
+        if (partition != null && partition < 0) {
+            throw new IllegalArgumentException(
+                    String.format("Invalid partition given with record: %d. Partition number should always greater than 0.", partition));
+        }
+
         // add topic to metadata topic list if it is not there already and reset expiry
         this.metadata.add(topic);
         Cluster cluster = metadata.fetch();
-        if (cluster.partitionsForTopic(topic) != null)
+        List<PartitionInfo> partitions = cluster.partitionsForTopic(topic);
+        // Return cached metadata if we have it, and if the record's partition is either undefined
+        // or within the known partition range
+        if (partitions != null && (partition == null || partition < partitions.size()))
```

Comment:
Please add a comment explaining why we refresh metadata only once if the partition is missing. I think it's really important to document non-obvious details like this so that both the intent and reason are clear.